apps/call.go

```diff
@@ -7,72 +7,106 @@ import (
 	"github.com/mattermost/mattermost-plugin-apps/server/utils/md"
 )
 
+// CallType determines what action is expected of a function.
 type CallType string
 
 const (
 	// CallTypeSubmit (default) indicates the intent to take action.
-	CallTypeSubmit = CallType("submit")
+	CallTypeSubmit CallType = "submit"
+
 	// CallTypeForm retrieves the form definition for the current set of values,
 	// and the context.
-	CallTypeForm = CallType("form")
+	CallTypeForm CallType = "form"
+
 	// CallTypeCancel is used for for the (rare?) case of when the form with
 	// SubmitOnCancel set is dismissed by the user.
-	CallTypeCancel = CallType("cancel")
+	CallTypeCancel CallType = "cancel"
+
 	// CallTypeLookup is used to fetch items for dynamic select elements
-	CallTypeLookup = CallType("lookup")
+	CallTypeLookup CallType = "lookup"
 )
 
-// A Call invocation is supplied a BotAccessToken as part of the context. If a
-// call needs acting user's or admin tokens, it should be specified in the
-// Expand section.
-//
-// If a user or admin token are required and are not available from previous
-// consent, the appropriate OAuth flow is launched, and the Call is executed
-// upon its success.
-//
-// TODO: what if a call needs a token and it was not provided? Return a call to
-// itself with Expand.
+// Call defines a way to invoke an App's function. Calls are used to fetch App's
+// bindings, to process notifications, and to respond to user input from forms,
+// and command line.
 type Call struct {
-	Path   string      `json:"path,omitempty"`
-	Expand *Expand     `json:"expand,omitempty"`
-	State  interface{} `json:"state,omitempty"`
+	// The path of the Call. For HTTP apps, the path is appended to the app's
+	// RootURL. For AWS Lambda apps, it is mapped to the appropriate Lambda name
+	// to invoke, and then passed in the call request.
+	Path string `json:"path,omitempty"`
+
+	// Expand specifies what extended data should be provided to the function in
+	// each request. It may be various auth tokens, configuration data, or
+	// details of Mattermost entities such as the acting user, current team and
+	// channel, etc.
```

Comment:
Where should we talk about default expands?

Reply:
I think we should expand by default the post on post actions. Apart of that, not sure if we should expand by default any other.

Reply:
> expand by default the post on post actions.

Why? And why for V1? We have now an easy pattern for creating default Call attributes for certain calls (OnInstall, Bindings) - we could use it, but I want to be wary of special cases.